pkg/dogstatsd/listeners/ratelimit/mem_based_rate_limiter.go

```diff
@@ -0,0 +1,172 @@
+// When the low memory limit is reached, Wait() blocks once and may try to release memory.
+// The memory limits are defined as soft limits.
+//
+// `memoryRateLimiter` provides a way to dynamically update the rate at which the memory limit
+// is checked.
+// When the soft limit is reached, we would like to wait and release memory but not block too often.
+// `freeOSMemoryRateLimiter` provides a way to dynamically update the rate at which `FreeOSMemory` is
+// called when the soft limit is reached.
+type MemBasedRateLimiter struct {
+	telemetry               telemetry
+	memoryUsage             memoryUsage
+	lowSoftLimitRate        float64
+	highSoftLimitRate       float64
+	memoryRateLimiter       *geometricRateLimiter
+	freeOSMemoryRateLimiter *geometricRateLimiter
+	previousMemoryUsageRate float64
+}
+
+type memoryUsage interface {
+	getMemoryUsageRate() (float64, error)
+}
+
+var memBasedRateLimiterTml = newMemBasedRateLimiterTelemetry()
+
+// BuildMemBasedRateLimiter builds a new instance of *MemBasedRateLimiter
+func BuildMemBasedRateLimiter() (*MemBasedRateLimiter, error) {
+	var memoryUsage memoryUsage
+	var err error
+	if memoryUsage, err = newCgroupMemoryUsage(); err == nil {
+		log.Info("cgroup limits detected")
+	} else {
+		memoryUsage = newHostMemoryUsage()
+		log.Infof("cgroup limits not detected")
+		log.Debugf("cgroup limits not detected: %v", err)
+	}
+
+	return NewMemBasedRateLimiter(
+		memBasedRateLimiterTml,
+		memoryUsage,
+		getConfigFloat("low_soft_limit"),
+		getConfigFloat("high_soft_limit"),
+		config.Datadog.GetInt("dogstatsd_mem_based_rate_limiter.go_gc"),
+		geometricRateLimiterConfig{
+			getConfigFloat("rate_check.min"),
+			getConfigFloat("rate_check.max"),
+			getConfigFloat("rate_check.factor")},
+		geometricRateLimiterConfig{
+			getConfigFloat("soft_limit_freeos_check.min"),
+			getConfigFloat("soft_limit_freeos_check.max"),
+			getConfigFloat("soft_limit_freeos_check.factor"),
+		},
+	)
+}
+
+func getConfigFloat(subkey string) float64 {
+	return config.Datadog.GetFloat64("dogstatsd_mem_based_rate_limiter." + subkey)
+}
+
+// NewMemBasedRateLimiter creates a new instance of MemBasedRateLimiter.
+func NewMemBasedRateLimiter(
+	telemetry telemetry,
+	memoryUsage memoryUsage,
+	lowSoftLimitRate float64,
+	highSoftLimitRate float64,
+	goGC int,
+	memoryRateLimiter geometricRateLimiterConfig,
+	freeOSMemoryRateLimiter geometricRateLimiterConfig) (*MemBasedRateLimiter, error) {
+
+	// When `SetMemoryLimit` will be available (https://github.com/golang/go/issues/48409),
+	//  SetGCPercent, madvdontneed=1 and debug.FreeOSMemory() can be removed.
+	if goGC > 0 {
+		debug.SetGCPercent(goGC)
+	}
+
+	// madvdontneed=1 forces the memory to be released. This allows the RSS to reflect what
+	// is really used by the Agent
+	if !strings.Contains(os.Getenv("GODEBUG"), "madvdontneed=1") {
+		return nil, errors.New("GODEBUG must have `madvdontneed=1` in order to use this feature")
+	}
```

Comment:
I believe this was only the case through go1.15?

Reply:
Code updated.